### DISCUSSION TITLE
Request: Support Implicit Expansion with Arithmetic Operations

### DISCUSSION THREAD

**Discussion by VestOfHolding** [optimatika/ojAlgo#531] [General, ojAlgo and related code]
**Is your feature request related to a problem? Please describe.**
When doing arithmetic operations on two matrices that aren't exactly the same size, but still compatible, I would like an option for the arithmetic operation to move forward anyway.

**Describe the solution you'd like**
MathWorks has a feature called[ implicit expansion](https://www.mathworks.com/help/coder/ug/generate-code-with-implicit-expansion-enabled.html) that I believe would be really useful here as well.

**Comment by apete** [comment]
I assume you want operations on matrices/arrays like this: https://se.mathworks.com/help/matlab/matlab_prog/compatible-array-sizes-for-basic-operations.html

Some variant of that could be supported. Any suggestions on how to implement this in ojAlgo?

**Comment by VestOfHolding** [comment]
Yes, that's correct.

I'm still kind of new to all this, so I'm not sure at the moment. I just noticed that this was a really neat feature that MathWorks/Matlab has that none of the active Java matrix libraries I've seen have, so it would be a cool way to stand out if added. Though on the page I linked, I do notice it talks about some of the code it generates when it realizes it needs to do the implicit expansion, if any ideas can be gained from that.

EDIT: Realized I missed half a sentence from my original post, so I've fixed that.

**Comment by apete** [comment]
I've added some support for this in a branch named [Compatible-Array-Sizes](https://github.com/optimatika/ojAlgo/tree/Compatible-Array-Sizes) that I may include in a future version.

There are new methods `fillCompatile(...)`, `modifyCompatile(...)` and `onCompatile(...)` to complement the already existing "fill", "modify" and "on" methods.

The factory interfaces got methods to construct instances of compatible sizes/shapes.

Vector space method like "add" and "subtract" have been redefined to no longer throw exceptions if dimensions are not equal, but instead broadcast/repeat rows or columns.
